File: plugin/src/main/java/de/infolektuell/gradle/jextract/GradleJextractPlugin.java
package de.infolektuell.gradle.jextract;

import de.infolektuell.gradle.jextract.extensions.JextractExtension;
import de.infolektuell.gradle.jextract.extensions.SourceSetExtension;
import de.infolektuell.gradle.jextract.service.JextractStore;
import de.infolektuell.gradle.jextract.tasks.*;

import static de.infolektuell.gradle.jextract.tasks.JextractBaseTask.*;

import org.gradle.api.*;
import org.gradle.api.file.Directory;
import org.gradle.api.file.FileCollection;
import org.gradle.api.plugins.JavaPlugin;
import org.gradle.api.plugins.JavaPluginExtension;
import org.gradle.api.provider.Provider;
import org.gradle.api.tasks.TaskProvider;
import org.gradle.internal.jvm.Jvm;
import org.gradle.jvm.toolchain.JavaLanguageVersion;
import org.jspecify.annotations.NonNull;

import java.util.*;

/// A Gradle plugin to add Jextract to the build
public abstract class GradleJextractPlugin implements Plugin<@NonNull Project> {
    /// The plugin ID that must be used in build scripts to apply the plugin
    public static final String PLUGIN_NAME = "de.infolektuell.jextract";

    ///  The gradle property to set a local Jextract installation path
    public static final String JEXTRACT_LOCAL_INSTALLATION_PROPERTY = "org.openjdk.jextract.installation-path";

    /// Used by Gradle
    public GradleJextractPlugin() { super(); }

    /// Configures the plugin if it is applied
    public void apply(Project project) {
        final JextractExtension extension = project.getObjects().newInstance(JextractExtension.class);
        project.getExtensions().add(JextractExtension.EXTENSION_NAME, extension);
        extension.getOutput().convention(project.getLayout().getBuildDirectory().dir("generated/sources/jextract/java"));
        extension.getGenerateSourceFiles().convention(false);
        extension.getInstallation().getDistributions().convention(extension.getDistributions());
        extension.getLibraries().configureEach(lib -> {
            lib.getLibraries().convention(List.of(lib.getName()));
            lib.getUseSystemLoadLibrary().convention(false);
            lib.getOutput().convention(extension.getOutput().dir(lib.getName()));
            lib.getGenerateSourceFiles().convention(extension.getGenerateSourceFiles());
        });

        project.getGradle().getSharedServices().registerIfAbsent(JextractStore.SERVICE_NAME, JextractStore.class, s -> {
            s.getMaxParallelUsages().convention(1);
            s.parameters(parameters -> {
                parameters.getCacheDir().convention(project.getRootProject().getLayout().getProjectDirectory().dir(".gradle/jextract"));
                parameters.getDistributions().convention(extension.getInstallation().getDistributions());
            });
        });

        project.getPlugins().apply(JavaPlugin.class);
        project.getPluginManager().withPlugin("java", javaPlugin -> {
            final JavaPluginExtension javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
            final Provider<@NonNull JavaLanguageVersion> javaVersion = javaExtension.getToolchain().getLanguageVersion()
                    .orElse(JavaLanguageVersion.of(Objects.requireNonNullElse(Jvm.current().getJavaVersionMajor(), 25)));
            extension.getInstallation().getJavaLanguageVersion().convention(javaVersion);

            final var localInstallationPath = project.getProviders().gradleProperty(JEXTRACT_LOCAL_INSTALLATION_PROPERTY);
            if (localInstallationPath.isPresent()) {
                final Directory location = project.getLayout().getProjectDirectory().dir(localInstallationPath.get());
                if (!location.getAsFile().isDirectory()) throw new GradleException(String.format("The path %s doesn't point to a directory.", location));
                extension.getInstallation().getLocation().convention(location);
            }
            final Provider<@NonNull JextractInstallation> jextractInstallation = extension.getInstallation().getLocation().map(location -> {
                    final var installation = project.getObjects().newInstance(LocalJextractInstallation.class);
                    installation.getLocation().convention(location);
                    return (JextractInstallation) installation;
                })
                .orElse(extension.getInstallation().getJavaLanguageVersion().map(version -> {
                    final var installation = project.getObjects().newInstance(RemoteJextractInstallation.class);
                    installation.getJavaLanguageVersion().convention(version);
                    return (JextractInstallation) installation;
                }));

            extension.getLibraries().configureEach(lib -> {
                project.getTasks().register(lib.getGenerateBindingsTaskName(), JextractGenerateTask.class, task -> {
                    task.setDescription("Uses Jextract to generate Java bindings for the " + lib.getName() + " native library");
                    task.getInstallation().convention(jextractInstallation);
                    task.getHeader().convention(lib.getHeader());
                    task.getIncludes().convention(lib.getIncludes());
                    task.getDefinedMacros().convention(lib.getDefinedMacros());
                    task.getHeaderClassName().convention(lib.getHeaderClassName());
                    task.getTargetPackage().convention(lib.getTargetPackage());
                    final Provider<@NonNull Map<@NonNull String, @NonNull Set<@NonNull String>>> whitelist = project.getProviders().provider(() -> {
                        return Map.of(
                            "function", lib.getWhitelist().getFunctions().get(),
                            "constant", lib.getWhitelist().getConstants().get(),
                            "struct", lib.getWhitelist().getStructs().get(),
                            "union", lib.getWhitelist().getUnions().get(),
                            "typedef", lib.getWhitelist().getTypedefs().get(),
                            "var", lib.getWhitelist().getVariables().get()
                        );
                    });
                    task.getWhitelist().convention(whitelist);
                    task.getArgFile().convention(lib.getWhitelist().getArgFile());
                    task.getLibraries().convention(lib.getLibraries());
                    task.getUseSystemLoadLibrary().convention(lib.getUseSystemLoadLibrary());
                    task.getGenerateSourceFiles().convention(lib.getGenerateSourceFiles());
                    task.getSources().convention(lib.getOutput());
                    task.getClasses().convention(project.getLayout().getBuildDirectory().dir("generated/classes/jextract/java/" + lib.getName()));
                });

                project.getTasks().register(lib.getDumpIncludesTaskName(), JextractDumpIncludesTask.class, task -> {
                    task.setDescription("Uses Jextract to dump all includes of the " + lib.getName() + " native library into an arg file");
                    task.getInstallation().convention(jextractInstallation);
                    task.getHeader().convention(lib.getHeader());
                    task.getIncludes().convention(lib.getIncludes());
                    task.getArgFile().convention(project.getLayout().getBuildDirectory().file("reports/jextract/" + lib.getName() + "-includes.txt"));
                });
            });

            javaExtension.getSourceSets().configureEach(s -> {
                final SourceSetExtension sourceSetExtension = project.getObjects().newInstance(SourceSetExtension.class);
                s.getExtensions().add(SourceSetExtension.EXTENSION_NAME, sourceSetExtension);
                sourceSetExtension.getLibraries().all(lib -> {
                    final TaskProvider<@NonNull JextractGenerateTask> task = project.getTasks().named(lib.getGenerateBindingsTaskName(), JextractGenerateTask.class);
                    s.getJava().srcDir(task.flatMap(JextractGenerateTask::getSources));
                    s.getResources().srcDir(task.flatMap(JextractGenerateTask::getSources));
                    final FileCollection classes = project.getLayout().files(task.flatMap(JextractGenerateTask::getClasses));
                    s.setCompileClasspath(s.getCompileClasspath().plus(classes));
                    s.setRuntimeClasspath(s.getRuntimeClasspath().plus(classes));
                });
            });
        });
    }
}

File: plugin/src/main/kotlin/de/infolektuell/gradle/jextract/GradleJextractPlugin.kt
package de.infolektuell.gradle.jextract

import de.infolektuell.gradle.jextract.extensions.JextractExtension
import de.infolektuell.gradle.jextract.extensions.SourceSetExtension
import de.infolektuell.gradle.jextract.service.JextractStore
import de.infolektuell.gradle.jextract.tasks.JextractBaseTask
import de.infolektuell.gradle.jextract.tasks.JextractDumpIncludesTask
import de.infolektuell.gradle.jextract.tasks.JextractGenerateTask
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.plugins.JavaPluginExtension
import org.gradle.api.tasks.SourceSetContainer
import org.gradle.api.tasks.TaskProvider
import org.gradle.internal.jvm.Jvm
import org.gradle.jvm.toolchain.JavaLanguageVersion

abstract class GradleJextractPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.gradle.sharedServices.registerIfAbsent(JextractStore.SERVICE_NAME, JextractStore::class.java) { s ->
            s.maxParallelUsages.convention(1)
            s.parameters.cacheDir.convention(project.rootProject.layout.projectDirectory.dir(".gradle/jextract"))
        }

        val extension = project.extensions.create(JextractExtension.EXTENSION_NAME, JextractExtension::class.java)
        val defaultInstallation = project.objects.newInstance(JextractBaseTask.RemoteJextractInstallation::class.java).apply {
            javaLanguageVersion.convention(JavaLanguageVersion.of(Jvm.current().javaVersionMajor ?: 22))
        }
        extension.installation.convention(defaultInstallation)
        extension.output.convention(project.layout.buildDirectory.dir("generated/sources/jextract"))
        extension.generateSourceFiles.convention(false)

        extension.libraries.configureEach { lib ->
            lib.useSystemLoadLibrary.convention(false)
            lib.output.convention(extension.output.dir(lib.name))
            lib.generateSourceFiles.convention(extension.generateSourceFiles)
        }

        val jextractGenerateTasks = mutableMapOf<String, TaskProvider<JextractGenerateTask>>()
        extension.libraries.all { lib ->
            val generateTaskProvider = project.tasks.register("${lib.name}JextractGenerateBindings", JextractGenerateTask::class.java) { task ->
                task.description = "Uses Jextract to generate Java bindings for the ${lib.name} native library"
                task.installation.set(extension.installation)
                task.header.set(lib.header)
                task.includes.set(lib.includes)
                task.definedMacros.set(lib.definedMacros)
                task.headerClassName.set(lib.headerClassName)
                task.targetPackage.set(lib.targetPackage)
                task.whitelist.run {
                    put("function", lib.whitelist.functions)
                    put("constant", lib.whitelist.constants)
                    put("struct", lib.whitelist.structs)
                    put("union", lib.whitelist.unions)
                    put("typedef", lib.whitelist.typedefs)
                    put("var", lib.whitelist.variables)
                }
                task.argFile.set(lib.whitelist.argFile)
                task.libraries.set(lib.libraries)
                task.useSystemLoadLibrary.set(lib.useSystemLoadLibrary)
                task.generateSourceFiles.set(lib.generateSourceFiles)
                task.sources.set(lib.output)
            }
            jextractGenerateTasks[lib.name] = generateTaskProvider
            project.tasks.register("${lib.name}JextractDumpIncludes", JextractDumpIncludesTask::class.java) { task ->
                task.description = "Uses Jextract to dump all includes of the ${lib.name} native library into an arg file"
                task.installation.set(extension.installation)
                task.header.set(lib.header)
                task.includes.set(lib.includes)
                task.argFile.set(project.layout.buildDirectory.file("reports/jextract/${lib.name}-includes.txt"))
            }
        }

        project.pluginManager.withPlugin("java") {
            project.extensions.findByType(JavaPluginExtension::class.java)?.let { defaultInstallation.javaLanguageVersion.convention(it.toolchain.languageVersion) }
            project.extensions.findByType(SourceSetContainer::class.java)?.all { s ->
                val sourceSetExtension = s.extensions.create(SourceSetExtension.EXTENSION_NAME, SourceSetExtension::class.java, project.objects)
                sourceSetExtension.libraries.all { lib ->
                    jextractGenerateTasks[lib.name]?.flatMap { it.sources }?.also {
                        s.java.srcDir(it)
                        s.resources.srcDir(it)
                        s.compileClasspath += project.files(it)
                        s.runtimeClasspath += project.files(it)
                    }
                }
            }
        }
    }

    companion object {
        const val PLUGIN_NAME = "de.infolektuell.jextract"
    }
}
